crypto/pkcs11key/key.go

```diff
@@ -357,14 +531,21 @@ func (ps *Key) Sign(rand io.Reader, msg []byte, opts crypto.SignerOpts) (signatu
 	}
 
 	// Add DigestInfo prefix
-	// TODO: Switch mechanisms based on CKA_KEY_TYPE
-	mechanism := []*pkcs11.Mechanism{pkcs11.NewMechanism(pkcs11.CKM_RSA_PKCS, nil)}
-	prefix, ok := hashPrefixes[hash]
-	if !ok {
-		err = errors.New("unknown hash function")
-		return
+	var mechanism []*pkcs11.Mechanism
+	var signatureInput []byte
+
+	if ps.keyType == RSA {
+		mechanism = []*pkcs11.Mechanism{pkcs11.NewMechanism(pkcs11.CKM_RSA_PKCS, nil)}
+		prefix, ok := hashPrefixes[hash]
```

Comment:
The hashPrefixes is a table of pre-encoded ASN.1 prefix for DigestInfo objects used in RSA PKCS#1 1.5 signature (see src/pkg/crypto/rsa/pkcs1v15.go). CKM_ECDSA operates on raw hash values.